Model/Sync/Catalog/Processor.php

```diff
@@ -175,8 +175,9 @@ public function process($forceSyncProducts = [], $storeIds = [])
                     }
                     $this->retryItems[$storeId] = [];
                     if ($this->productSyncLimit > 0) {
-                        $forceSyncProductIds = $forceSyncProducts[$storeId] ?? $forceSyncProducts;
-                        $collection = $this->getCollectionForSync($forceSyncProductIds);
+                        $forceSyncProductIds = $this->isCommandLineSync &&
```

Comment:
@manaftryzens 
we don't understand the fix.
The only process that sends products per storeId is the command line.
There's a case where a product ID is the same as a storeId, so it will refer to the same "position" in the collection. But this would raise an exception anyway, and also not was the intent of this fix.

Reply:
During commandline sync, we are preparing an array of products in the below format

[storeId1 => [productId1, productId2, productId3], storeId2 => productId4, product5, product6]

This time this is an associative multi dimensional array with storeID as it's keys and productIds are inner arrays

If there are 2 store views with ID = 1 & 2, the array will form like below.

[1 => [productId1, productId2, productId3], 2 => productId4, product5, product6]

This array is passed to the process() method.

When it reaches at this line at , 

foreach ($allStores as $storeId) (This loop is there at the beginning of the function) {
...
....
....
....
$forceSyncProductIds = $forceSyncProducts[$storeId] ?? $forceSyncProducts;
.....
.....
....


}

When the loop is on storeId=1 , the array [productId1, productId2, productId3] will be passed to the getCollectionForSync() method and it will process the sync for these particular products only.

All good so far and the commandline flow works very well as expected.

Let's consider a scenario, where we try to sync an order that contains some products in it which are not yet synced.

During order sync we will have an array in the below format, that contains the products to be synced as part of that order.

$unSyncedProductIds = $this->getUnSyncedProductIds($productIds, $visibleItems, $storeId);
$sync = $this->process($unSyncedProductIds, [$storeId]);

This time it is passing $unSyncedProductIds as an array  in the below format 

[productId1, productId2, productId3]

This time this is a normal array. In PHP a normal array will have hidden keys associated with it (Sorry, please ignore if you already know this) . So in reality this array will get the below format

[0 => productId1, 1 => productId2, 2 => productId3]

Now when it reaches at the below line

foreach ($allStores as $storeId) (This loop is there at the beginning of the function) {
...
....
....
....
$forceSyncProductIds = $forceSyncProducts[$storeId] ?? $forceSyncProducts;
.....
.....
....
}

When the loop is on store ID=1 , productId2 alone will be passed to the getCollectionForSync() method

Now if you check getCollectionForSync($forceSyncProductIds) method, it has the line below.

if ($unSyncedProductIds && is_array($unSyncedProductIds)) {
    $collection->addFieldToFilter('entity_id', ['in' => $unSyncedProductIds]);
}

As $unSyncedProductIds is not an array this time (it is a single value as productId2), it will not hit inside the above if condition. 

As a result, this time it will collect all the products from the DB and try to sync it by ignoring what we actually required to do.

If my assumption is correct, this can be one reason why order cron is stuck as it is trying to sync all the products each time when it try to sync a single order with some missing products in it.

Reply:
@manaftryzens 
Thanks Manaf for looking into the orders cron stuck issue. 
now the fix makes more sense. 

I’m afraid that the root cause is not the missing conditions and checks that you added, but it is because the first argument is sometimes an array and sometimes a map/hash. 

The issue would have been avoided if the argument was following a single type. 

As part of the successful path to a beta and code stabilization please follow the map “typed” argument and adjust the orders code to send a
[storeId => [p1, p2]]

Reply:
Thanks @manaftryzens